docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,171 @@
+
+* The rust components themselves are designed to be consumed as individual components - the "logins" component doesn't know anything about the "bookmarks" component.
+
+There are a couple of gotchyas in the current implementations too - there's an issue when certain engines don't yet appear in meta/global - see bug 1479929 for all the details.
+
+The tl;dr of the above is that each rust component should be capable of working with different sync managers. In all cases, it is important that meta/global management is done in a holistic and consistent manner.
+
+# Approach for the rust components
+
+In general, we need to pass certain state around, between the "manager" and the engines. We must define this in a way so that the manager can be implemented anywhere - we should not assume the manager is implemented in rust.
+
+We also need to consider the "clients" engine in this state - we will need to communicate commands sent by other devices and commands that we wish to send to other devices.
+
+We will rely on the consuming app to ensure that the sync manager is initialized correctly before syncing any engines.
+
+While we can "trust" each engine, we should try, as much as possible, to isolate engine data from other engines. For example, the keys used to encrypt a collection should only be exposed to that specific engine, and there's no need for one engine to know what info/collections returns for other engines.
+
+# Implementation plan
+
+* The sync manager is responsible for persisting the global syncID and version, considering the state to be "new" if it differs, and failing fatally if the version isn't acceptable. Engines never need to see these values.
+
+* The sync manager is responsible for persisting and populating the "declined" engine list. If an engine is declined, the manager should not ask it to sync. IOW, an engine doesn't really need to know whether it is enabled or not. However, engine should expose a way to be "reset" when the manager notices the declined state for an engine has been changed. This will allow engines to optimize certain things, such as not keeping tomstones etc. As part of this, the sync manager must expose an API so the consuming application can change the disabled engine state.
+
+* The sync manager is responsible for managing the "clients" collection - engines should never see details about that engine. However, engines will need the ability to be told about commands targetting that engine. The sync manager will also need to be able to accept commands to be delivered to other clients - although that needs more thought as the conditions in which a command need to be sent typically aren't managed by the engine itself - eg, when a "bookmark" restore is done.
+
+* Each engine is responsible for persisting engine-specific stuff which needs to be persisted. An obvious thing which needs to be persisted is the syncID and version string for the engine itself, so it can take the necessary action when these change.
```

Comment:
I'd prefer if the sync manager persisted the complete `m/g`. Each engine should also store the `syncID`, or some way to tie it to the current state, but we'll run the risk of uploading an incomplete `m/g` if we rely on engines to persist the data. (For example, what if we had an `m/g` entry for `addresses`, but we don't support the addresses engine?)

Reply:
Hmm, though...if we need to reupload `m/g`, say, because an engine's sync ID changes, we could always fetch it from the server, change _just that engine_ in the record, and upload, leaving the other engines alone.

Reply:
I've expanded on this a little